packages/vite/src/node/preview.ts

```diff
@@ -48,12 +48,28 @@ export async function preview(
 
   const options = config.server
   const hostname = resolveHostname(serverOptions.host ?? options.host)
-  const port = serverOptions.port ?? 5000
+  let port = serverOptions.port ?? 5000
   const protocol = options.https ? 'https' : 'http'
   const logger = config.logger
   const base = config.base
 
+  const onError = (e: Error & { code?: string }) => {
+    if (e.code === 'EADDRINUSE') {
+      logger.info(`Port ${port} is in use, trying another one...`)
+      httpServer.listen(++port, hostname.host)
```

Comment:
**question:** So this will recursive call the `onError` until it found one (or crash if it reaches 65536 😆)?

Reply:
I refer to `vite dev`:blush:
https://github.com/vitejs/vite/blob/0f45c218d195c37ad6f3f3b4a660bf071bb59a4a/packages/vite/src/node/server/index.ts#L595-L607

If there are too many retries, the user should be able to manually interrupt.

Reply:
Oh... If this already exists for the dev server, should we extract this into a util/shared?